src/main/java/org/commonjava/util/partyline/FileTree.java

```diff
@@ -327,46 +327,55 @@ boolean tryLock( File file, String ownerName, String label, LockLevel lockLevel,
         while ( end < 1 || System.currentTimeMillis() < end )
         {
             T result = tryLock( realFile, Thread.currentThread().getName(), "Open File for " + ( doOutput ? "output" : "input" ),
-                                           doOutput ? LockLevel.write : read, WAIT_TIMEOUT, unit, ( opLock ) -> {
+                                           doOutput ? LockLevel.write : read, timeout, unit, ( opLock ) -> {
                         FileEntry entry = entryMap.get( realFile.getAbsolutePath() );
-                        if ( entry.file != null )
+                        boolean proceed = false;
+                        synchronized ( entry )
```

Comment:
When two threads are going to collide and try to manipulate the same FileEntry at the same time, the map should always have a lock instance available. The first one in will create it, and before it can get GC'ed (because the first thread still hasn't finished) the second thread will try to use the file, find the lock in the map, and have to wait. That's the intention.

In general, I agree that depending on GC for functionality isn't good. I meant that I was depending on GC to contain the size of that map. In this case, the only time GC can remove a lock is if the thread that used the string and the lock is done. By that time, the collision will be over...maybe.

Actually, thinking about that more, that's not true. JoinableFile maintains a reference to the lock, and if the one in the map gets removed they will fall out of sync. So, if we don't use a WeakHashMap to contain the operation locks, then we'll need a way to ensure they get cleaned up, I guess. Ordinarily, I'd say we could do this in the FileCallbacks.closed() method, called by JoinableFile.close(). However in this case if another thread is waiting on the ReentrantLock trying to start working on the same file, then removing from the map when JoinableFile.close() is called isn't going to help.

I'm worried about leaving this in a ConcurrentHashMap without some method of cleaning up...it'll be a slow memory leak.